File: crates/examples/caller_demo.py
#!/usr/bin/env python3
"""
Caller / Callable Demo — RPC pattern

Demonstrates:
1. Callable server with inline handler
2. Caller making RPC calls
3. JSON request/response
"""

import asyncio
import json

from brainos.bus import BusManager
from brainos.caller import Caller, Callable


async def demo_inline_handler():
    print("═" * 60)
    print("  Demo 1 — Callable with inline handler")
    print("═" * 60)

    async with BusManager() as bus:
        def add_handler(req: str) -> str:
            a, b = map(int, req.split(","))
            return str(a + b)

        srv = await Callable.create(bus.bus, "rpc/add", add_handler)
        await srv.start()

        caller = await Caller.create(bus.bus, "rpc/add")
        result = await caller.call_text("5,7")
        print(f"  📤 Call: '5,7'")
        print(f"  📥 Response: '{result}'")
        print(f"  ✅ Inline handler done\n")


async def demo_run_handler():
    print("═" * 60)
    print("  Demo 2 — Callable with run() handler")
    print("═" * 60)

    async with BusManager() as bus:
        def multiply(req: str) -> str:
            a, b = map(int, req.split(","))
            return str(a * b)

        srv = await Callable.create(bus.bus, "rpc/mul")

        async def run_srv():
            try:
                await srv.run(multiply)
            except asyncio.CancelledError:
                pass

        task = asyncio.create_task(run_srv())
        await asyncio.sleep(0.2)

        caller = await Caller.create(bus.bus, "rpc/mul")
        result = await caller.call_text("6,7")
        print(f"  📤 Call: '6,7'")
        print(f"  📥 Response: '{result}'")
        print(f"  ✅ Run handler done\n")

        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            pass


async def demo_json_rpc():
    print("═" * 60)
    print("  Demo 3 — Callable with JSON handler")
    print("═" * 60)

    async with BusManager() as bus:
        def greet_handler(data):
            name = data.get("name", "World")
            return {"greeting": f"Hello, {name}!", "status": "ok"}

        srv = await Callable.create(bus.bus, "rpc/greet")

        async def run_srv():
            try:
                await srv.run_json(greet_handler)
            except asyncio.CancelledError:
                pass

        task = asyncio.create_task(run_srv())
        await asyncio.sleep(0.2)

        caller = await Caller.create(bus.bus, "rpc/greet")
        result = await caller.call_text(json.dumps({"name": "Alice"}))
        resp = json.loads(result)
        print(f"  📤 Call: {{\"name\": \"Alice\"}}")
        print(f"  📥 Response: {resp}")
        print(f"  ✅ JSON RPC done\n")

        task.cancel()
        try:
            await task
        except asyncio.CancelledError:
            pass


async def main():
    print("\n" + "📞" * 30)
    print("  BrainOS — Caller / Callable Demo")
    print("📞" * 30 + "\n")

    await demo_inline_handler()
    await demo_run_handler()
    await demo_json_rpc()

    print("═" * 60)
    print("  ✅ All Caller/Callable demos completed!")
    print("═" * 60 + "\n")


if __name__ == "__main__":
    asyncio.run(main())

File: crates/jsbos/examples/caller_demo.js
#!/usr/bin/env node
/**
 * Caller / Callable Demo — RPC pattern
 *
 * Demonstrates:
 * 1. Callable server with inline handler
 * 2. Callable with run() handler
 * 3. JSON request/response
 *
 * Usage:
 *     node crates/jsbos/examples/caller_demo.js
 */

const { Bus, version } = require('../jsbos.js');

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function demoInlineHandler() {
  console.log('═'.repeat(60));
  console.log('  Demo 1 — Callable with inline handler');
  console.log('═'.repeat(60));

  const bus = await Bus.create();
  console.log('  🚌 Bus created');

  const callable = await bus.createCallable('rpc/add');
  await callable.start();
  console.log('  📡 Callable started: rpc/add');

  await sleep(200);

  const caller = await bus.createCaller('rpc/add');
  const result = await caller.callText('5,7');
  console.log(`  📤 Call: '5,7'`);
  console.log(`  📥 Response: '${result}'`);
  console.log('  ✅ Inline handler done\n');
}

async function demoRunHandler() {
  console.log('═'.repeat(60));
  console.log('  Demo 2 — Callable with run() handler');
  console.log('═'.repeat(60));

  const bus = await Bus.create();
  console.log('  🚌 Bus created');

  const callable = await bus.createCallable('rpc/mul');
  callable.setHandler((err, input) => {
    const [a, b] = input.split(',').map(Number);
    return String(a * b);
  });
  await callable.start();
  console.log('  📡 Callable started: rpc/mul (custom handler: a*b)');

  await sleep(200);

  const caller = await bus.createCaller('rpc/mul');
  const result = await caller.callText('6,7');
  console.log(`  📤 Call: '6,7'`);
  console.log(`  📥 Response: '${result}'`);
  console.log('  ✅ Run handler done\n');
}

async function demoJsonRpc() {
  console.log('═'.repeat(60));
  console.log('  Demo 3 — Callable with JSON handler');
  console.log('═'.repeat(60));

  const bus = await Bus.create();
  console.log('  🚌 Bus created');

  const callable = await bus.createCallable('rpc/greet');
  callable.setHandler((err, input) => {
    const data = JSON.parse(input);
    const name = data.name || 'World';
    return JSON.stringify({ greeting: `Hello, ${name}!`, status: 'ok' });
  });
  await callable.start();
  console.log('  📡 Callable started: rpc/greet (JSON handler)');

  await sleep(200);

  const caller = await bus.createCaller('rpc/greet');
  const result = await caller.callText(JSON.stringify({ name: 'Alice' }));
  const resp = JSON.parse(result);
  console.log(`  📤 Call: {"name": "Alice"}`);
  console.log(`  📥 Response: ${JSON.stringify(resp)}`);
  console.log('  ✅ JSON RPC done\n');
}

async function main() {
  console.log('\n' + '📞'.repeat(30));
  console.log(`  BrainOS (jsbos v${version()}) — Caller / Callable Demo`);
  console.log('📞'.repeat(30) + '\n');

  await demoInlineHandler();
  await demoRunHandler();
  await demoJsonRpc();

  console.log('═'.repeat(60));
  console.log('  ✅ All Caller/Callable demos completed!');
  console.log('═'.repeat(60) + '\n');
}

main().catch(console.error).finally(() => process.exit(0));
